apps/full-site-editing/full-site-editing-plugin/block-patterns/class-block-patterns.php

```diff
@@ -41,27 +43,32 @@ public static function get_instance() {
 	}
 
 	/**
-	 * Extends block editor settings to include FSE block patterns.
-	 *
-	 * @param array $settings Default editor settings.
-	 * @return array Filtered editor settings.
+	 * Register FSE block patterns and categories.
 	 */
-	public function register_patterns( $settings ) {
-		// Remove core patterns except 'Two Columns of Text'.
-		$settings['__experimentalBlockPatterns'] = wp_list_filter(
-			$settings['__experimentalBlockPatterns'],
-			array(
-				'title' => 'Two Columns of Text',
-			)
-		);
+	public function register_patterns() {
+		// Register Block Pattern Categories.
+		if ( class_exists( 'WP_Block_Pattern_Categories_Registry' ) ) {
+			register_block_pattern_category( 'blog', array( 'label' => _x( 'Blog', 'Block pattern category', 'gutenberg' ) ) );
+			register_block_pattern_category( 'call-to-action', array( 'label' => _x( 'Call to Action', 'Block pattern category', 'gutenberg' ) ) );
+			register_block_pattern_category( 'contact', array( 'label' => _x( 'Contact', 'Block pattern category', 'gutenberg' ) ) );
+			register_block_pattern_category( 'images', array( 'label' => _x( 'Images', 'Block pattern category', 'gutenberg' ) ) );
+			register_block_pattern_category( 'menu', array( 'label' => _x( 'Menu', 'Block pattern category', 'gutenberg' ) ) );
+		}
 
-		// Add our patterns.
-		$settings['__experimentalBlockPatterns'] = array_merge(
-			$this->get_patterns(),
-			$settings['__experimentalBlockPatterns']
-		);
+		if ( class_exists( 'WP_Block_Patterns_Registry' ) && ! \WP_Block_Patterns_Registry::get_instance()->is_registered( 'text-two-columns' ) ) {
```

Comment:
It looks like the `class_exists` check currently guards the entire module. Alternatively you could update/replace [the existing one in the loader file](https://github.com/Automattic/wp-calypso/blob/5f5d7be03cc15aa509082f9c561f7f784967b81c/apps/full-site-editing/full-site-editing-plugin/full-site-editing-plugin.php#L245) and keep this from being loaded in the first place. Not super important but thought I'd share.

Can you help me understand the purpose of the `is_registered` check?

Reply:
> It looks like the `class_exists` check currently guards the entire module. Alternatively you could update/replace [the existing one in the loader file](https://github.com/Automattic/wp-calypso/blob/5f5d7be03cc15aa509082f9c561f7f784967b81c/apps/full-site-editing/full-site-editing-plugin/full-site-editing-plugin.php#L245) and keep this from being loaded in the first place. Not super important but thought I'd share.

Yeah, good point, thanks! I _think_ I like to have a check for the existence of these methods colocated in the module (class, even) that uses them, so a consumer of that class won't accidentally load it and break if they aren't available :thinking:

> Can you help me understand the purpose of the `is_registered` check?

Moar copy-pasta TBH:
https://github.com/WordPress/gutenberg/blob/d09cbf93c0250111b60a71cecf69e5c28c50563b/lib/client-assets.php#L696

The `is_registered` check _there_ is probably because of this:

https://github.com/WordPress/gutenberg/blob/d09cbf93c0250111b60a71cecf69e5c28c50563b/lib/client-assets.php#L692-L694

But that doesn't really apply to us (none of "our" patterns are potentially registerd by Core), so it should be  fine to remove :+1:

Reply:
532888e